### PR TITLE
Fix several package manager issues

### LINKED ISSUE #2282: remote: left over msgs (eg pkg mgr) need to be cleared
repro:

open pkg mgr
connect to remote
pkg mgr now displays:

image

### LINKED ISSUE #2357: Package manager stuck
None of the tabs are working, they are all stuck with the spinning "Loading" message.  I'm connect to local Microsoft R Client.

Logs here:
\\pytools\r\Bugs\PackageMgrStuck

### LINKED ISSUE #2510: Unable to load installed package in remote
image

### PR DESCRIPTION
**Pull request by AlexanderSher**
- Fix #2357: Package manager stuck
- Fix #2510: Unable to load installed package in remote
- Fix #2282: remote: left over msgs (eg pkg mgr) need to be cleared

### CODE REVIEW

File: src/Common/Core/Impl/Events/DirtyEventSource.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        private readonly object _source;
+        private int _dirty;
+
+        public event EventHandler Event;
+
+        public DirtyEventSource(object source) {
+            _source = source;
+        }
+
+        public bool IsDirty {
```
Comment (MikhailArkhipov) [inline src/Common/Core/Impl/Events/DirtyEventSource.cs]: We don't seem to be using `get` so maybe just `Reset` and `FireOnce` ?

File: src/R/Components/Impl/PackageManager/Implementation/ViewModel/RPackageManagerViewModel.cs [hunk 2]
```diff
@@ -307,55 +306,70 @@ internal class RPackageManagerViewModel : BindableBase, IRPackageManagerViewMode
...
-        private void AddErrorMessage(string message) {
+        private void AddErrorMessage(string message, ErrorMessageType messageType) {
             _coreShell.AssertIsOnMainThread();
             if (FirstError == null) {
                 FirstError = message;
+                _firstErrorType = messageType;
             } else {
-                _errorMessages.Enqueue(message);
+                _errorMessages.Add(new ErrorMessage(message, messageType));
                 HasMultipleErrors = true;
```
Comment (MikhailArkhipov) [inline src/R/Components/Impl/PackageManager/Implementation/ViewModel/RPackageManagerViewModel.cs]: Can it driven from count of messages in the list?

File: src/R/Components/Impl/PackageManager/Implementation/ViewModel/RPackageManagerViewModel.cs [hunk 3]
```diff
@@ -307,55 +306,70 @@ internal class RPackageManagerViewModel : BindableBase, IRPackageManagerViewMode
...
+            _errorMessages.RemoveWhere(e => e.Type == messageType);
+            HasMultipleErrors = _errorMessages.Count > 0;
+
+            if (_firstErrorType == messageType) {
+                DismissErrorMessage();
+            }
+        }
 
         public void DismissAllErrorMessages() {
             _coreShell.AssertIsOnMainThread();
```
Comment (MikhailArkhipov) [inline src/R/Components/Impl/PackageManager/Implementation/ViewModel/RPackageManagerViewModel.cs]: Maybe just make it concurrent and not worry about main thread?

File: src/Common/Core/Impl/Events/DirtyEventSource.cs (hunk 1 above)
Comment (AlexanderSher) [inline src/Common/Core/Impl/Events/DirtyEventSource.cs]: Sure

File: src/R/Components/Impl/PackageManager/Implementation/ViewModel/RPackageManagerViewModel.cs (hunk 3 above)
Comment (AlexanderSher) [inline src/R/Components/Impl/PackageManager/Implementation/ViewModel/RPackageManagerViewModel.cs]: Actually no, collection isn't bound.

File: src/R/Components/Impl/PackageManager/Implementation/ViewModel/RPackageManagerViewModel.cs [hunk 4]
```diff
@@ -24,43 +23,42 @@
...
     internal class RPackageManagerViewModel : BindableBase, IRPackageManagerViewModel {
         private readonly IRPackageManager _packageManager;
-        private readonly IRSession _session;
         private readonly IRSettings _settings;
         private readonly ICoreShell _coreShell;
         private readonly BinaryAsyncLock _availableLock;
         private readonly BinaryAsyncLock _installedAndLoadedLock;
         private readonly BatchObservableCollection<object> _items;
-        private readonly Queue<string> _errorMessages;
+        private readonly List<ErrorMessage> _errorMessages;
```
Comment (AlexanderSher) [inline src/R/Components/Impl/PackageManager/Implementation/ViewModel/RPackageManagerViewModel.cs]: Ok, let me try it.

File: src/R/Components/Impl/PackageManager/Implementation/ViewModel/RPackageManagerViewModel.cs (hunk 3 above)
Comment (MikhailArkhipov) [inline src/R/Components/Impl/PackageManager/Implementation/ViewModel/RPackageManagerViewModel.cs]: Some of them appear to be bound to the collection state such as HasMultipleErrors which can be readonly props

**Comment by MikhailArkhipov** [comment]
👍

**Pull request merged**